neuralmonkey/decoders/transformer.py

```diff
@@ -0,0 +1,334 @@
+"""Implementation of the decoder of the Transformer model.
+
+Described in Vaswani et al. (2017), arxiv.org/abs/1706.03762
+"""
+from typing import Callable, Set, List, Tuple  # pylint: disable=unused-import
+
+import tensorflow as tf
+from typeguard import check_argument_types
+
+from neuralmonkey.attention.scaled_dot_product import (
+    attention, empty_multi_head_loop_state)
+from neuralmonkey.attention.base_attention import (
+    Attendable, get_attention_states, get_attention_mask)
+from neuralmonkey.decorators import tensor
+from neuralmonkey.decoders.autoregressive import (
+    AutoregressiveDecoder, LoopState, extend_namedtuple, DecoderHistories,
+    DecoderFeedables)
+from neuralmonkey.encoders.transformer import (
+    TransformerLayer, position_signal)
+from neuralmonkey.logging import log
+from neuralmonkey.nn.utils import dropout
+from neuralmonkey.vocabulary import (
+    Vocabulary, PAD_TOKEN_INDEX, END_TOKEN_INDEX)
+
+# pylint: disable=invalid-name
+TransformerHistories = extend_namedtuple(
+    "RNNHistories",
+    DecoderHistories,
+    [("decoded_symbols", tf.TensorArray),
+     ("self_attention_histories", List[Tuple]),
+     ("inter_attention_histories", List[Tuple]),
+     ("input_mask", tf.TensorArray)])
+# pylint: enable=invalid-name
+
+
+class TransformerDecoder(AutoregressiveDecoder):
+
+    # pylint: disable=too-many-arguments
+    def __init__(self,
+                 name: str,
+                 encoder: Attendable,
+                 vocabulary: Vocabulary,
+                 data_id: str,
+                 # TODO infer the default for these three from the encoder
+                 ff_hidden_size: int,
+                 n_heads_self: int,
+                 n_heads_enc: int,
+                 depth: int,
+                 max_output_len: int = None,
+                 dropout_keep_prob: float = 1.0,
+                 attention_dropout_keep_prob: float = 1.0,
+                 save_checkpoint: str = None,
+                 load_checkpoint: str = None) -> None:
+        check_argument_types()
+        AutoregressiveDecoder.__init__(
+            self,
+            name=name,
+            vocabulary=vocabulary,
+            data_id=data_id,
+            max_output_len=max_output_len,
+            dropout_keep_prob=dropout_keep_prob,
+            # from the "dark slide"
```

Comment:
tamto někam zapadlo tak to tady opakuju - tenhle koment vyhoď a label smoothing dej do argumentů konstruktoru - chceme to štelovat z konfigu